File: iroha-balancer-core/src/test/java/soramitsu/irohautils/balancer/IntegrationTests.java
package soramitsu.irohautils.balancer;

import com.codahale.metrics.Counter;
import com.fasterxml.jackson.databind.ObjectMapper;
import iroha.protocol.Endpoint;
import iroha.protocol.TransactionOuterClass;
import jp.co.soramitsu.crypto.ed25519.Ed25519Sha3;
import jp.co.soramitsu.iroha.java.Transaction;
import jp.co.soramitsu.iroha.java.Utils;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.apache.camel.CamelContext;
import org.apache.camel.ProducerTemplate;
import org.apache.camel.component.metrics.MetricsComponent;
import org.apache.camel.component.rabbitmq.RabbitMQConstants;
import org.junit.jupiter.api.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.reactive.AutoConfigureWebTestClient;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ContextConfiguration;
import org.testcontainers.junit.jupiter.Testcontainers;
import soramitsu.irohautils.balancer.client.config.RMQConfig;
import soramitsu.irohautils.balancer.service.IrohaService;

import javax.xml.bind.DatatypeConverter;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.SortedMap;

import static iroha.protocol.Endpoint.TxStatus.COMMITTED;
import static jp.co.soramitsu.iroha.testcontainers.detail.GenesisBlockBuilder.*;

@Testcontainers
@SpringBootTest
@ContextConfiguration(initializers = TestContainersMock.Initializer.class)
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
@AutoConfigureWebTestClient
@Slf4j
public class IntegrationTests {

    public static final String RABBITMQ_TRANSACTIONS_PRODUCER =
            "rabbitmq:iroha-balancer?exchangeType=topic&durable=true&autoDelete=false&skipQueueDeclare=true";

    public static final String TORII_ROUTING_KEY = "torii";
    public static final String LIST_TORII_ROUTING_KEY = "list-torii";
    public static final Ed25519Sha3 crypto = new Ed25519Sha3();

    @Autowired
    private IrohaService irohaService;
    @Autowired
    private ProducerTemplate producerTemplate;
    @Autowired
    private CamelContext camelContext;
    @Autowired
    private ObjectMapper objectMapper;

    @Order(10)
    @Test
    @Timeout(30)
    void createAccountsTest() throws Exception {
        List<byte[]> txs = new ArrayList<>();
        List<String> hashes = new ArrayList<>();
        int totalTx = 1000;
        for (int i = 0; i < totalTx; i++) {
            TransactionOuterClass.Transaction createAccountTransaction = Transaction.builder(defaultAccountId)
                    .createAccount("account_" + i, defaultDomainName, crypto.generateKeypair().getPublic())
                    .build()
                    .sign(defaultKeyPair)
                    .build();
            txs.add(createAccountTransaction.toByteArray());
            hashes.add(Utils.toHexHash(createAccountTransaction));
        }
        txs.forEach(tx -> {
            producerTemplate.send(RABBITMQ_TRANSACTIONS_PRODUCER, exchange -> {
                exchange.getMessage().setBody(tx);
                exchange.getMessage().setHeader(RabbitMQConstants.ROUTING_KEY, TORII_ROUTING_KEY);
            });
        });
        hashes.parallelStream().forEach(this::checkCommitted);

        SortedMap<String, Counter> metrics = ((MetricsComponent) camelContext.getComponent("metrics")).getMetricRegistry().getCounters();
        metrics.values().forEach(metric ->
                Assertions.assertTrue(metric.getCount() > (totalTx * .9)/irohaService.getIrohaPeers().size()));

    }

    @Order(20)
    @Test
    @Timeout(30)
    void batchTransactionTest() throws Exception {
        Transaction transaction1 = Transaction.builder(defaultAccountId)
                .createAsset("usd", defaultDomainName, 2)
                .build();
        Transaction transaction2 = Transaction.builder(defaultAccountId)
                .addAssetQuantity("usd#" + defaultDomainName, "1000")
                .build();
        List<String> hashes = Arrays.asList(DatatypeConverter.printHexBinary(Utils.reducedHash(transaction1)),
                DatatypeConverter.printHexBinary(Utils.reducedHash(transaction2)));
        TransactionOuterClass.Transaction tx1 = transaction1.makeMutable()
                .setBatchMeta(TransactionOuterClass.Transaction.Payload.BatchMeta.BatchType.ATOMIC, hashes)
                .build()
                .sign(defaultKeyPair)
                .build();
        TransactionOuterClass.Transaction tx2 = transaction2.makeMutable()
                .setBatchMeta(TransactionOuterClass.Transaction.Payload.BatchMeta.BatchType.ATOMIC, hashes)
                .build()
                .sign(defaultKeyPair)
                .build();
        byte[][] txs = new byte[][]{tx1.toByteArray(), tx2.toByteArray()};

        producerTemplate.send(RABBITMQ_TRANSACTIONS_PRODUCER, exchange -> {
            exchange.getMessage().setBody(objectMapper.writeValueAsString(txs));
            exchange.getMessage().setHeader(RabbitMQConstants.ROUTING_KEY, LIST_TORII_ROUTING_KEY);
        });
        checkCommitted(Utils.toHexHash(tx1));
        checkCommitted(Utils.toHexHash(tx2));
    }

    private void checkCommitted(String trxHash1) {
        Endpoint.TxStatus txStatus = getTxStatus(trxHash1);
        while (COMMITTED != txStatus) {
            try {
                Thread.sleep(500);
            } catch (InterruptedException e) {
            }
            txStatus = getTxStatus(trxHash1);
            log.info("Status for {}: {}", trxHash1, txStatus);

        }
    }

    private Endpoint.TxStatus getTxStatus(String trxHash1) {
        return irohaService.getIrohaPeers().values().stream()
                .findAny().get()
                .txStatusSync(DatatypeConverter.parseHexBinary(trxHash1))
                .getTxStatus();
    }

}

File: iroha-balancer-integration-test/src/test/kotlin/soramitsu/iroha/balancer/integration_test/IntegrationTests.kt
package soramitsu.iroha.balancer.integration_test

import iroha.protocol.Endpoint.TxStatus
import iroha.protocol.TransactionOuterClass
import jp.co.soramitsu.crypto.ed25519.Ed25519Sha3
import jp.co.soramitsu.crypto.ed25519.Utils
import jp.co.soramitsu.iroha.java.IrohaAPI
import jp.co.soramitsu.iroha.java.Transaction
import jp.co.soramitsu.iroha.java.Utils.toHexHash
import jp.co.soramitsu.iroha.testcontainers.detail.GenesisBlockBuilder.*
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ContextConfiguration
import org.testcontainers.containers.Network
import org.testcontainers.junit.jupiter.Testcontainers
import soramitsu.iroha.balancer.integration_test.TestContainersMock.network
import soramitsu.irohautils.balancer.client.config.RMQConfig
import soramitsu.irohautils.balancer.client.service.IrohaBalancerClientService
import soramitsu.irohautils.balancer.helper.ContainerHelper
import javax.xml.bind.DatatypeConverter

@Testcontainers
@SpringBootTest
@ContextConfiguration(initializers = [TestContainersMock.Initializer::class])
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class IntegrationTests {

    private val containerHelper = ContainerHelper()

    private val irohaBalancerCoreContextFolder = "${containerHelper.userDir}/../iroha-balancer-core/build/docker/"
    private val irohaBalancerCoreDockerfile = "${containerHelper.userDir}/../iroha-balancer-core/build/docker/Dockerfile"

    private val irohaBalancerCoreContainer =
            containerHelper.createIrohaBalancerCoreContainer(irohaBalancerCoreContextFolder, irohaBalancerCoreDockerfile)
                    .withNetwork(Network.SHARED)

    val crypto = Ed25519Sha3()
    private lateinit var client: IrohaBalancerClientService

    @BeforeAll
    fun init() {
        irohaBalancerCoreContainer
                .withEnv("CAMEL_COMPONENT_RABBITMQ_HOSTNAME", TestContainersMock.rabbitmq.host)
                .withEnv("CAMEL_COMPONENT_RABBITMQ_PORT_NUMBER", TestContainersMock.rabbitmq.getMappedPort(5672).toString())
                .withEnv("CAMEL_COMPONENT_RABBITMQ_USERNAME", "guest")
                .withEnv("CAMEL_COMPONENT_RABBITMQ_PASSWORD", "guest")
                .start()

        val rmqConfig = RMQConfig(
                TestContainersMock.rabbitmq.host,
                TestContainersMock.rabbitmq.getMappedPort(5672),
                "guest",
                "guest"
        )
        client = IrohaBalancerClientService(rmqConfig)
    }

    @Test
    fun dummy() {
        val transaction: TransactionOuterClass.Transaction = Transaction.builder(defaultAccountId)
                .createAccount("account_", defaultDomainName, crypto.generateKeypair().getPublic())
                .build()
                .sign(defaultKeyPair)
                .build()
        client.balanceToTorii(transaction)

        val iroha = IrohaAPI(network.toriiAddresses[0])
        checkCommitted(toHexHash(transaction), iroha)
    }
    private fun checkCommitted(trxHash1: String, iroha: IrohaAPI) {
        var txStatus = iroha.txStatusSync(trxHash1.toByteArray()).txStatus
        while (TxStatus.COMMITTED != txStatus) {
            try {
                Thread.sleep(500)
            } catch (e: InterruptedException) {
            }
            txStatus = iroha.txStatusSync(trxHash1.toByteArray()).txStatus
            println("Status for $trxHash1: $txStatus")
        }
    }
}
